File: evaluation/S0/python/Medium/1581-move-pieces-to-obtain-a-string.py
# Source: https://leetcode.com/problems/move-pieces-to-obtain-a-string/   |   Difficulty: Medium
#
# Problem Description:
# You are given two strings start and target, both of length n. Each string consists only of the characters 'L', 'R', and '_' where:
#
#
# 	The characters 'L' and 'R' represent pieces, where a piece 'L' can move to the left only if there is a blank space directly to its left, and a piece 'R' can move to the right only if there is a blank space directly to its right.
# 	The character '_' represents a blank space that can be occupied by any of the 'L' or 'R' pieces.
#
#
# Return true if it is possible to obtain the string target by moving the pieces of the string start any number of times. Otherwise, return false.
#
# Example:
# Input: start = "_L__R__R_", target = "L______RR"
# Output: true
# Explanation: We can obtain the string target from start by doing the following moves:
# - Move the first piece one step to the left, start becomes equal to "L___R__R_".
# - Move the last piece one step to the right, start becomes equal to "L___R___R".
# - Move the second piece three steps to the right, start becomes equal to "L______RR".
# Since it is possible to get the string target from start, we return true.
#
# Constraints:
# n == start.length == target.length
# 	1 <= n <= 105
# 	start and target consist of the characters 'L', 'R', and '_'.
#

class Solution(object):
    def canChange(self, start, target):
        """
        :type start: str
        :type target: str
        :rtype: bool
        """
        s_ptr, t_ptr = 0, 0
        n = len(start)
        
        while s_ptr < n or t_ptr < n:
            while s_ptr < n and start[s_ptr] == '_':
                s_ptr += 1
            while t_ptr < n and target[t_ptr] == '_':
                t_ptr += 1
            
            if s_ptr == n and t_ptr == n:
                return True
            if s_ptr == n or t_ptr == n:
                return False
            
            if start[s_ptr] != target[t_ptr]:
                return False
            
            if start[s_ptr] == 'L' and s_ptr < t_ptr:
                return False
            if start[s_ptr] == 'R' and s_ptr > t_ptr:
                return False
            
            s_ptr += 1
            t_ptr += 1
        
        return True

File: evaluation/S3/c/Medium/1581-move-pieces-to-obtain-a-string.c
// Source: https://leetcode.com/problems/move-pieces-to-obtain-a-string/   |   Difficulty: Medium
//
// Problem Description:
// You are given two strings start and target, both of length n. Each string consists only of the characters 'L', 'R', and '_' where:
//
//
// 	The characters 'L' and 'R' represent pieces, where a piece 'L' can move to the left only if there is a blank space directly to its left, and a piece 'R' can move to the right only if there is a blank space directly to its right.
// 	The character '_' represents a blank space that can be occupied by any of the 'L' or 'R' pieces.
//
//
// Return true if it is possible to obtain the string target by moving the pieces of the string start any number of times. Otherwise, return false.
//
// Example:
// Input: start = "_L__R__R_", target = "L______RR"
// Output: true
// Explanation: We can obtain the string target from start by doing the following moves:
// - Move the first piece one step to the left, start becomes equal to "L___R__R_".
// - Move the last piece one step to the right, start becomes equal to "L___R___R".
// - Move the second piece three steps to the right, start becomes equal to "L______RR".
// Since it is possible to get the string target from start, we return true.
//
// Constraints:
// n == start.length == target.length
// 	1 <= n <= 105
// 	start and target consist of the characters 'L', 'R', and '_'.
//
// Helpful references (internal KB):
// - Z-function and its calculation (string, array, two-pointers)
//   • When to use: Use when you need to find the length of the longest common prefix between a string and all its suffixes efficiently. It is fundamental for various string matching and pattern searching problems.
//   • Idea: The Z-function computes an array where each element z[i] is the length of the longest common prefix between the original string and its suffix starting at index i. This algorithm calculates the entire Z-array in O(N) time complexity.
//   • Invariants: For all k < i, z[k] has been correctly computed.; The interval [l, r) always represents the rightmost segment match found so far.
//   • Tips: Leverage previously computed Z-values to avoid redundant comparisons.; Maintain a [l, r) window representing the rightmost segment match found so far.
//   • Pitfalls: Incorrectly handling the base case z[0] (usually set to 0).; Off-by-one errors when calculating the length of the prefix or updating the window.
// - Suffix Array (array, string, simulation, two-pointers)
//   • When to use: Use when needing to efficiently compare substrings, find all occurrences of a pattern, or determine lexicographical order among suffixes within a string.
//   • Idea: A Suffix Array stores sorted suffixes of a string, enabling fast substring comparisons and pattern searching. Construction typically takes O(N log N) time, with comparisons often O(1) after preprocessing.
//   • Invariants: The suffix array `p` contains all suffixes of the string `s` sorted lexicographically.; During construction, equivalence classes `c` correctly group suffixes that are identical up to a certain length `2^k`.
//   • Tips: Precompute log values for efficient block length calculation.; Utilize equivalence classes for O(1) comparison of power-of-two length substrings.
//   • Pitfalls: Complexity of suffix array construction can be tricky (e.g., O(N log N) vs O(N)).; Off-by-one errors when calculating block indices or lengths.
// - Substring Search (string, array, two-pointers, simulation)
//   • When to use: Use KMP to efficiently find all occurrences of a pattern string within a larger text. It is particularly useful when the pattern might appear multiple times or when avoiding redundant text scans is crucial for performance.
//   • Idea: The Knuth-Morris-Pratt (KMP) algorithm efficiently finds all occurrences of a pattern string within a text string by pre-processing the pattern to avoid re-scanning characters in the text. It achieves a time complexity of O(N + M), where N is text length and M is pattern length.
//   • Invariants: The `next` array is correctly pre-computed such that `next[k]` indicates the longest proper prefix of `pattern[0...k-1]` that is also a suffix of `pattern[0...k-1]`.; During the search phase, `text[i-j ... i-1]` always matches `pattern[0 ... j-1]` if `j > 0`.
//   • Tips: Pre-compute the 'next' array (also known as LPS or failure array) for the pattern.; The 'next' array stores the length of the longest proper prefix of the pattern that is also a suffix of the pattern up to that index.
//   • Pitfalls: Incorrectly building the 'next' (LPS) array, especially handling the base cases or transitions.; Off-by-one errors in array indexing for either the text or pattern pointers.
//
// Notes:
// - Keep the given function signature unchanged.
// - Use the C standard library only; no external packages.
// - Do not change input/output formats or return value semantics.
// - Conformance: ISO C (C89/C99/C11). Do NOT use GNU C extensions (nested functions, statement expressions, typeof, __auto_type, etc.). Avoid VLAs and non-standard builtins.
// - If you introduce any helper functions, include their full implementations in this file; no hidden helpers.

bool canChange(char* start, char* target) {
    int i = 0, j = 0;
    while (start[i] != '\0' && target[j] != '\0') {
        while (start[i] == '_') i++;
        while (target[j] == '_') j++;
        
        if (start[i] == '\0' || target[j] == '\0') break;
        
        if (start[i] != target[j]) return false;
        
        if (start[i] == 'L' && i < j) return false;
        if (start[i] == 'R' && i > j) return false;
        
        i++;
        j++;
    }
    
    while (start[i] == '_') i++;
    while (target[j] == '_') j++;
    
    return start[i] == '\0' && target[j] == '\0';       
}
